Implement `workbench.fileDialog.homePath` setting (fix #115730)

gjsjohnmurray opened the pull request:
This PR resolves #115730 by adding a machine-level setting `workbench.fileDialog.homePath` that will override the user home folder, which is what the file and folder dialogs currently fall back to when no better choice is derived.

gjsjohnmurray commented:
@bpasero any chance you could take a look at this? I don't think it's controversial, and reckon it will please a lot of people.

bpasero commented:
Since this impacts simple file dialogs as well, adding @alexr00 for a look.

@gjsjohnmurray some high level feedback:
* we have an existing setting `files.simpleDialog.enable` so maybe this setting should be called `files.dialog.homePath`
* there is another [`getUserHome()`](https://github.com/microsoft/vscode/blob/a201c730802a9e27f038b1c95fa70a284d91c93b/src/vs/workbench/services/dialogs/browser/simpleFileDialog.ts#L251) method that probably needs to adopt the setting too
* I find the code labeled as `Normalize homePath setting so simple file dialog works correctly` strange, why is `this.pathService.fileURI` not sufficient alone?

alexr00 commented:
> I find the code labeled as `Normalize homePath setting so simple file dialog works correctly` strange, why is this.pathService.fileURI not sufficient alone?

+1 I would like to understand this too.

gjsjohnmurray commented:
> > I find the code labeled as `Normalize homePath setting so simple file dialog works correctly` strange, why is this.pathService.fileURI not sufficient alone?
> 
> +1 I would like to understand this too.

IIRC the https://github.com/microsoft/vscode/pull/182908/commits/59c93233224cc54f47c29e4eced8b75842556136 commit was required to fix test failure. I'll investigate exactly which (the relevant CI log has already been purged).

gjsjohnmurray commented:
I had misremembered. I added this when testing how my change behaved in the remote scenario, based on this advice:

https://github.com/microsoft/vscode/issues/162874#issuecomment-1271774905

When debugging Code - OSS I used this command to open a remote:

![image](https://github.com/microsoft/vscode/assets/6726799/ee5221b3-71f9-4f59-a0ca-be20d1cc8e6d)

Then I used Settings Editor to set a value for my setting (in screenshot below I have already renamed it as suggested by @bpasero ):

![image](https://github.com/microsoft/vscode/assets/6726799/349e9c2b-6f3e-43f4-82a0-fb7f31a74bcc)

As my workstation is Windows, and the TestResolver remote treats my Windows filesystem as though on a remote, the intuitive format for the value I set is a Windows path.

But in the preferredHome() method this PR adds, userHomePromise resolves to a Uri whose _formatted property is `'vscode-remote://test%2Btest/c%3A/Users/JohnM'` and calling its `with` method with the argument `{ path: dialogHomePath }` throws:
```
[UriError]: If a URI contains an authority component, then the path component must either be empty or begin with a slash ("/") character
```
because I had set homePath to `C:\Users\JohnM\Documents\GitHub`

I'm not currently catching this, so the dialog never opens and the user doesn't get told why.

alexr00 commented on src/vs/workbench/contrib/files/browser/files.contribution.ts:
```diff
@@ -323,6 +323,11 @@ configurationRegistry.registerConfiguration({
 		},
+		'files.dialog.homePath': {
+			'type': 'string',
+			'description': nls.localize('fileDialogHomePath', "Default path for file dialogs, overriding user's home path."),
```
It might be worth calling out in the description that default is first based on the last folder or file you had open and only then if that doesn't exist will this setting be uses.

gjsjohnmurray replied:
Good idea. I have pushed a change.

gjsjohnmurray replied:
I have also pushed a change so `~` replacement uses the true home.

alexr00 reviewed (changes requested):
Thank you for the details on `Normalize homePath setting so simple file dialog works correctly`. 

> there is another getUserHome() method that probably needs to adopt the setting too

`userHome` is used in 2 places in the simple file picker. I think we should only use the setting value in one of those places: 

https://github.com/gjsjohnmurray/vscode/blob/b7034b80526c5b8ca5c27e1419868919898fceaa/src/vs/workbench/services/dialogs/browser/simpleFileDialog.ts#L204-L209

The other place is about replacing `~` and I don't think we should override that ever.

alexr00 reviewed (dismissed):
Looks good to me!

bpasero commented on src/vs/workbench/services/dialogs/browser/abstractFileDialogService.ts:
```diff
@@ -86,12 +86,24 @@ export abstract class AbstractFileDialogService implements IFileDialogService {
+		}
+
+		// Normalize homePath setting so simple file dialog works correctly
+		const path = (await this.pathService.fileURI(dialogHomePath)).path;
```
I am sorry, but this change looks like a hack. It cannot go in like that.

gjsjohnmurray replied:
As I understand things, `preferredHome` must return the type (i.e. scheme) of URI that the caller requested (e.g. 'file' or 'vscode-remote'). And when I'm running through the debugger the `this.pathService.fileUri` method always returned a 'file://' URI regardless of whether I worked locally or remotely (using TestResolver).

I will push some more inline comments, plus a change to use `inspect` to retrieve the setting value because when working on a remote I don't think we should fall back to a local user-level path value if no remote one exists. @alexr00 might have a view about this.

bpasero replied:
Lets step back. You have a configured path that is not a URI and you want to convert that to the best URI for the environment you are in. That should only be 1 call and not 2. I suggest to use this:

https://github.com/microsoft/vscode/blob/ff3369f2329fb18601dcd1f51655cd5445535974/src/vs/base/common/resources.ts#L433

With the `pathService.defaultUriScheme`, e.g. like here:

https://github.com/microsoft/vscode/blob/ff3369f2329fb18601dcd1f51655cd5445535974/src/vs/workbench/services/textfile/browser/textFileService.ts#L587

This should make sure that the non-URI path gets converted to the best local URI matching the environment.

gjsjohnmurray replied:
Genius! I have pushed the change.

bpasero commented:
I just found this method in the file picker:

https://github.com/microsoft/vscode/blob/2bd04a1380e93fa92294878b17305f246f7f334a/src/vs/workbench/services/dialogs/browser/simpleFileDialog.ts#L222-L233

Which looks like it is doing a similar thing. Maybe @alexr00 could comment and maybe this needs to be used in this case here? It seems to do a bit more what I had suggested though.

bpasero commented on src/vs/workbench/contrib/files/browser/files.contribution.ts:
```diff
@@ -323,6 +323,11 @@ configurationRegistry.registerConfiguration({
 			'default': 'askUser',
 			'scope': ConfigurationScope.LANGUAGE_OVERRIDABLE
 		},
+		'files.dialog.homePath': {
```
I wonder if the `default` of this setting should be filled in based on the `IPathService.userHome`

gjsjohnmurray replied:
@bpasero I'm unclear what you're suggesting. Currently if the user leaves this machine-level string setting unset or sets it to "" in JSON they will continue getting the existing behaviour.

bpasero replied:
I think its fine. My idea was to show to the user the default path that we pick based on the environment.

alexr00 commented:
> Which looks like it is doing a similar thing. Maybe @alexr00 could comment and maybe this needs to be used in this case here? It seems to do a bit more what I had suggested though.

I don't think that full method is needed here. The simple file picker has to do two extra things that the setting doesn't need to:
- Display uniform slashes (all `/` or all `\`)
- Switch between the remote file system and local file system.

bpasero commented:
@gjsjohnmurray any thoughts on the name of the setting, only now I realise its called `files.dialog.homePath`, but isn't this rather `files.dialog.defaultPath`?

gjsjohnmurray commented:
@bpasero I don't feel strongly about the current name. Fine with me if you want to rename it. Thanks for accepting the PR.

gjsjohnmurray commented:
I updated the title of this PR to reflect the setting rename in #186156
